core/services/workflows/v2/secrets.go

```diff
@@ -2,17 +2,244 @@ package v2
+	lggr.Debug("fetching secrets...")
+
+	resp, err := vaultCap.Execute(ctx, capabilities.CapabilityRequest{
+		Payload:      anypbReq,
+		Method:       vault.MethodGetSecrets,
+		CapabilityId: vault.CapabilityID,
+		Metadata: capabilities.RequestMetadata{
+			WorkflowOwner: s.workflowOwner,
+			WorkflowName:  s.workflowName,
+		},
+	})
+	if err != nil {
+		lggr.Errorw("failed to fetch secrets", "err", err)
+		return nil, fmt.Errorf("failed to execute vault.GetSecrets: %w", err)
+	}
+
+	lggr.Debug("successfully fetched secrets")
+
+	respPayload := &vault.GetSecretsResponse{}
+	err = resp.Payload.UnmarshalTo(respPayload)
+	if err != nil {
+		lggr.Errorw("failed to unmarshal vault payload to GetSecretsResponse", "err", err)
+		return nil, fmt.Errorf("failed to unmarshal vault payload to GetSecretsResponse: %w", err)
+	}
+
+	m := map[string]*vault.SecretResponse{}
+	for _, s := range respPayload.Responses {
+		key := keyFor(s.Owner, s.Namespace, s.Id)
+		m[key] = s
+	}
+
+	sdkResp := []*sdkpb.SecretResponse{}
+	for _, r := range request.Requests {
+		key := keyFor(s.workflowOwner, r.Namespace, r.Id)
+		resp, ok := m[key]
+		if !ok {
+			lggr.Debugw("could not find secret in response map", "key", key)
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     "could not find secret for " + key,
+					},
+				},
+			})
+			continue
+		}
+
+		if resp.Error != "" {
+			lggr.Debugw("secret request returned an error", "key", key, "err", resp.Error)
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     resp.Error,
+					},
+				},
+			})
+			continue
+		}
+
+		if len(resp.EncryptedDecryptionKeyShares) != 1 {
+			lggr.Errorw("unexpected number of decryption shares received", "key", key, "len", len(resp.EncryptedDecryptionKeyShares))
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     fmt.Sprintf("expected one set of decryption shares, got %d", len(resp.EncryptedDecryptionKeyShares)),
+					},
+				},
+			})
+			continue
+		}
+
+		shares := resp.EncryptedDecryptionKeyShares[0].Shares
+		secret, err := s.decrypter(shares)
+		if err != nil {
+			lggr.Errorw("failed to combine decryption shares", "key", key, "err", err)
+			sdkResp = append(sdkResp, &sdkpb.SecretResponse{
+				Response: &sdkpb.SecretResponse_Error{
+					Error: &sdkpb.SecretError{
+						Id:        r.Id,
+						Namespace: r.Namespace,
+						Owner:     s.workflowOwner,
+						Error:     err.Error(),
```

Comment:
I'm not sure if we should forward those low-level errors to the workflow. Maybe replace with a generic one?